File: JAVA/level2/탑.java
import java.util.*;

class Solution {
    public int[] solution(int[] heights) {
        int[] answer = new int[heights.length];
        Stack<Integer> stack = new Stack<Integer>();
        answer[0] = 0;
        stack.push(10000 + heights[0]);

        for(int i=1; i<heights.length; i++){
            Stack<Integer> temp = stack;

            while(!temp.empty()){
                if(heights[i] < stack.peek()%10000 ){
                    answer[i] = stack.peek()/10000;
                    break;
                }
                temp.pop();
            }

            if(temp.empty())
                answer[i] = 0;

            stack.push((i+1)*10000 + heights[i]);
        }

        return answer;
    }
}

File: C++/level2/탑.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> heights) {
    vector<int> answer;

    for(int i=0; i<heights.size(); i++){
        if(i == 0)
            answer.push_back(0);
        else{
            for(int j=i-1; j>=0; j--){
                if(heights[j] > heights[i]){
                    answer.push_back(j+1);
                    break;
                }
                if(j == 0  && heights.at(j) <= heights.at(i))
                    answer.push_back(0);
            }
        }
    }

    return answer;
}
